packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -1291,6 +1293,131 @@ Hardhat Network's forking functionality only works with blocks from at least spu
     );
   }
 
+  public async getFeeHistory(
+    blockCount: BN,
+    newestBlock: BN | "pending",
+    rewardPercentiles: number[]
+  ): Promise<FeeHistory> {
+    const latestBlock = await this.getLatestBlockNumber();
+    const pendingBlockNumber = latestBlock.addn(1);
+
+    const resolvedNewestBlock =
+      newestBlock === "pending" ? pendingBlockNumber : newestBlock;
+
+    const oldestBlock = BN.max(
+      resolvedNewestBlock.sub(blockCount).addn(1),
+      new BN(0)
+    );
+
+    const baseFeePerGas: BN[] = [];
+    const gasUsedRatio: number[] = [];
+    const reward: BN[][] = [];
+
+    const lastBlock = resolvedNewestBlock.addn(1);
+
+    // We get the pending block here, and only if necessary, as it's something
+    // constly to do.
+    let pendingBlock: Block | undefined;
+    if (lastBlock.gte(pendingBlockNumber)) {
+      pendingBlock = await this.getBlockByNumber("pending");
+    }
+
+    for (
+      let blockNumber = oldestBlock;
+      blockNumber.lte(lastBlock);
+      blockNumber = blockNumber.addn(1)
+    ) {
+      if (blockNumber.lt(pendingBlockNumber)) {
+        // We know the block exists
+        const block = (await this.getBlockByNumber(blockNumber))!;
+        baseFeePerGas.push(block.header.baseFeePerGas ?? new BN(0));
+
+        if (blockNumber.lt(lastBlock)) {
+          gasUsedRatio.push(this._getGasUsedRatio(block));
+
+          if (rewardPercentiles.length > 0) {
+            reward.push(await this._getRewards(block, rewardPercentiles));
+          }
+        }
+      } else if (blockNumber.eq(pendingBlockNumber)) {
+        // This can only be run with EIP-1559, so this exists
+        baseFeePerGas.push((await this.getNextBlockBaseFeePerGas())!);
+
+        if (blockNumber.lt(lastBlock)) {
+          gasUsedRatio.push(this._getGasUsedRatio(pendingBlock!));
+
+          if (rewardPercentiles.length > 0) {
+            // We don't compute this for the pending block, as there's no
+            // effective miner fee yet.
+            reward.push(rewardPercentiles.map((_) => new BN(0)));
+          }
+        }
+      } else if (blockNumber.eq(pendingBlockNumber.addn(1))) {
+        baseFeePerGas.push(pendingBlock!.header.calcNextBaseFee());
+      }
```

Comment:
I think it's worth adding an `else { assert(false) } ` here.

Reply:
Good idea